ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -179,30 +181,128 @@ public void onForkChoiceUpdatedResult(
+                            .onAttestation(transaction, getIndexedAttestation(attestation));
+                        transaction.commit();
+                      })
+                  .thenApply(__ -> validationResult);
+            })
+        .exceptionallyCompose(
+            error -> {
+              final Throwable rootCause = Throwables.getRootCause(error);
+              if (rootCause instanceof InvalidCheckpointException) {
+                return SafeFuture.completedFuture(
+                    AttestationProcessingResult.invalid(rootCause.getMessage()));
+              }
+              return SafeFuture.failedFuture(error);
+            });
+  }
+
+  public void applyIndexedAttestations(final List<ValidatableAttestation> attestations) {
+    onForkChoiceThread(
+            () -> {
+              final VoteUpdater transaction = recentChainData.startVoteUpdate();
+              final ForkChoiceStrategy forkChoiceStrategy = getForkChoiceStrategy();
+              attestations.stream()
+                  .map(this::getIndexedAttestation)
+                  .forEach(
+                      attestation -> forkChoiceStrategy.onAttestation(transaction, attestation));
+              transaction.commit();
+            })
+        .ifExceptionGetsHereRaiseABug();
+  }
+
+  public void onAttesterSlashing(
+      final AttesterSlashing slashing,
+      InternalValidationResult validationStatus,
+      boolean fromNetwork) {
+    if (!validationStatus.isAccept()) {
+      return;
+    }
+    onForkChoiceThread(
+            () -> {
+              final VoteUpdater transaction = recentChainData.startVoteUpdate();
+              storeEquivocatingIndices(slashing, transaction);
+              transaction.commit();
+            })
+        .ifExceptionGetsHereRaiseABug();
   }
 
-  public SafeFuture<Boolean> processHead(final UInt64 nodeSlot, final boolean isPreProposal) {
-    return processHead(Optional.of(nodeSlot), isPreProposal);
+  public void subscribeToOptimisticHeadChangesAndUpdate(OptimisticHeadSubscriber subscriber) {
+    optimisticSyncSubscribers.subscribe(subscriber);
+    getOptimisticSyncing().ifPresent(subscriber::onOptimisticHeadChanged);
+  }
+
+  public void onTick(
+      final UInt64 currentTimeMillis, final Optional<TickProcessingPerformance> performanceRecord) {
+    final UpdatableStore store = recentChainData.getStore();
+    final UInt64 slotAtStartOfTick = spec.getCurrentSlot(store);
+    tickProcessor.onTick(currentTimeMillis).join();
+    performanceRecord.ifPresent(TickProcessingPerformance::tickProcessorComplete);
+    final UInt64 currentSlot = spec.getCurrentSlot(store);
+    if (currentSlot.isGreaterThan(slotAtStartOfTick)) {
+      applyDeferredAttestations(currentSlot).ifExceptionGetsHereRaiseABug();
+    }
+    performanceRecord.ifPresent(TickProcessingPerformance::deferredAttestationsApplied);
+  }
+
+  private void initializeProtoArrayForkChoice() {
+    initialized = true;
```

Comment:
I'd put `volatile` here, to be super safe

Reply:
removed the idea anyway...